src/Mod/Arch/ArchRoof.py

```diff
@@ -42,241 +52,328 @@ def QT_TRANSLATE_NOOP(ctxt,txt):
         self.Type = "Roof"
 
-    def onDocumentRestored(self,obj):
-
-        ArchComponent.Component.onDocumentRestored(self,obj)
+    def onDocumentRestored(self, obj):
+        ArchComponent.Component.onDocumentRestored(self, obj)
         self.setProperties(obj)
 
-    def calcHeight(self, id):
-
-        " Get the height from run and angle of the given roof profile "
+    def flipEdges(self, edges):
+        edges.reverse()
+        newEdges = []
+        for edge in edges:
+            NewEdge = DraftGeomUtils.edg(edge.Vertexes[1].Point, edge.Vertexes[0].Point)
+            newEdges.append(NewEdge)
+        return newEdges
 
-        htRel = self.profilsDico[id]["run"]*(math.tan(math.radians(self.profilsDico[id]["angle"])))
+    def calcHeight(self, id):
+        '''Get the height from run and angle of the given roof profile'''
+        htRel = self.profilsDico[id]["run"] * (math.tan(math.radians(self.profilsDico[id]["angle"])))
         return htRel
 
     def calcRun(self, id):
-
-        " Get the run from height and angle of the given roof profile "
-
-        runRel = self.profilsDico[id]["height"]/(math.tan(math.radians(self.profilsDico[id]["angle"])))
+        '''Get the run from height and angle of the given roof profile'''
+        runRel = self.profilsDico[id]["height"] / (math.tan(math.radians(self.profilsDico[id]["angle"])))
         return runRel
 
     def calcAngle(self, id):
-
-        " Get the angle from height and run of the given roof profile "
-
-        a = math.degrees(math.atan(self.profilsDico[id]["height"]/self.profilsDico[id]["run"]))
-        return a
+        '''Get the angle from height and run of the given roof profile'''
+        ang = math.degrees(math.atan(self.profilsDico[id]["height"] / self.profilsDico[id]["run"]))
+        return ang
 
     def getPerpendicular(self, vec, rotEdge, l):
-
-
-        " Get the perpendicular vec of given edge on xy plane "
-        norm = FreeCAD.Vector(0,0,1)
-        if hasattr(self,"normal"):
+        '''Get the perpendicular vec of given edge on xy plane'''
+        norm = Vector(0.0, 0.0, 1.0)
+        if hasattr(self, "normal"):
             if self.normal:
-                norm = FreeCAD.Vector(self.normal)
-        perpendicular = vec.cross(norm)
-        if  -180. <= rotEdge < -90.:
-            perpendicular[0] = abs(perpendicular[0])*-1
-            perpendicular[1] = abs(perpendicular[1])*-1
-        elif   -90. <= rotEdge <= 0.:
-            perpendicular[0] = abs(perpendicular[0])*-1
-            perpendicular[1] = abs(perpendicular[1])
-        elif 0. < rotEdge <= 90.:
-            perpendicular[0] = abs(perpendicular[0])
-            perpendicular[1] = abs(perpendicular[1])
-        elif 90. < rotEdge <= 180.:
-            perpendicular[0] = abs(perpendicular[0])
-            perpendicular[1] = abs(perpendicular[1])*-1
+                norm = self.normal
+        per = vec.cross(norm)
```

Comment:
Remember that verbose is better than terse, so `perpendicular` is better than `per`.

The reason `self.normal` is passed through `Vector` is that `self.normal` may not be a vector, it may be a simple tuple `(0, 0, 1)`. So, I guess this code was written in this way to transform a tuple or list into an actual `Vector`, and be able to perform the cross product.

Maybe in the past `self.normal` was not guaranteed to be a `Vector` so this conversion was necessary. If you checked this is no longer the case, then removing the transformation is not a problem, but otherwise I would check that `self.normal` is assigned properly.

Reply:
See above.